site-src/geps/gep-724.md

```diff
@@ -0,0 +1,614 @@
+for Route->Service references, but is embedded within the Route and Gateway
+resources. The alternatives below explore what this could look like with
+ReferencePolicy.
+
+## API Changes
+
+The proposed changes here can be summarized as:
+
+* Remove Route selector from the RouteBindingSelector in Gateway listeners.
+* Replace Route kind and group with optional list of accepted kinds and groups
+  in RouteBindingSelector.
+* Rename RouteBindingSelector to ListenerRoutes.
+* Replace the 3 options from Route -> Gateway (All, FromList, SameNamespace)
+  with a reference list that supports arbitrary kinds.
+
+### Gateway
+
+On Gateway, the only change involves removing the Route selector field.
+Everything else remains the same.
+
+#### Removed
+
+```go
+  // Selector specifies a set of route labels used for selecting
+  // routes to associate with the Gateway. If this Selector is defined,
+  // only routes matching the Selector are associated with the Gateway.
+  // An empty Selector matches all routes.
+  //
+  // Support: Core
+  //
+  // +optional
+  Selector *metav1.LabelSelector `json:"selector,omitempty"`
+	// Group is the group of the route resource to select. Omitting the value or specifying
+	// the empty string indicates the networking.x-k8s.io API group.
+	// For example, use the following to select an HTTPRoute:
+	//
+	// routes:
+	//   kind: HTTPRoute
+	//
+	// Otherwise, if an alternative API group is desired, specify the desired
+	// group:
+	//
+	// routes:
+	//   group: acme.io
+	//   kind: FooRoute
+	//
+	// Support: Core
+	//
+	// +optional
+	// +kubebuilder:default=networking.x-k8s.io
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=253
+	Group *string `json:"group,omitempty"`
+	// Kind is the kind of the route resource to select.
+	//
+	// Kind MUST correspond to kinds of routes that are compatible with the
+	// application protocol specified in the Listener's Protocol field.
+	//
+	// If an implementation does not support or recognize this
+	// resource type, it SHOULD set the "ResolvedRefs" condition to false for
+	// this listener with the "InvalidRoutesRef" reason.
+	//
+	// Support: Core
+	Kind string `json:"kind"`
+```
+
+#### Added
+```go
+type ListenerRoutes struct {
+  // ...
+  // Kinds specifies the groups and kinds of Routes that are allowed to bind to
+  // this Gateway listener. When unspecified or empty, the only limitation on
+  // the kinds of Routes supported is the Listener protocol. Kind MUST
+  // correspond to kinds of Routes that are compatible with the application
+  // protocol specified in the Listener's Protocol field. If an implementation
+  // does not support or recognize this resource type, it SHOULD set the
+  // "ResolvedRefs" condition to false for this listener with the
+  // "InvalidRoutesRef" reason.
+  //
+  // Support: Core
+  //
+  // +optional
+  Kinds []RouteGroupKind `json:"kinds,omitempty"`
+}
+
+type RouteGroupKind struct {
```

Comment:
My preference would be to require both and then make them optional in future (not a breaking change).

Reply:
> My preference would be to require both and then make them optional in future

Agree with that. That's the current state as I understand it - even though Group is marked as "optional" that's only because there is a default value. There is no way for it to be empty.

Reply:
This means that the way these fields are used is that if this stanza is not specified, match anything. If it is, match only this. You must specify at least a Kind, in which case the group defaults to `gateway.networking.io`. Otherwise, you may specify any Kind.

I like it.